---
title: "DSeq2"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{DSeq2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Setup
```{r setup}
library(multiAnalysis)
library(tximport)
# library(rhdf5)
# Load matrix for treatment
set_analysis(block_name = "emna_treat", type = "qual")
# set_analysis(block_name = "still_manif", type = "qual")
# Extract variable for discrimination

load(file.path(path1, "clinic_transf.rda"))
batches <- clinic_transf$batch %>% set_names(clinic_transf$immun_aid_identifier)

load(file.path(path1, "clinic_transf.rda"))

if (!file.exists(file.path(path_img, "shrinked"))) {
    dir.create(file.path(path_img, "shrinked"))
}

setEnrichrSite("Enrichr")
# listEnrichrDbs()$libraryName[listEnrichrDbs()$libraryName %>% grepl("orphanet", ., ignore.case = TRUE)]
dbs <- c("GO_Molecular_Function_2021", "GO_Biological_Process_2021", "Orphanet_Augmented_2021", "WikiPathway_2021_Human", "MSigDB_Hallmark_2020", "Reactome_2022")
```

# Load gene count data
```{r}
# Load the count matrix
path <- file.path(golem::get_golem_wd(), "inst", "extdata")
processed_matrix <- read_tsv(file = file.path(path, "processed_gene_matrix.tsv"))
counts_matrix <- read_delim(file = file.path(path, "raw_gene_matrix.txt"), delim = " ", skip = 1, col_names = colnames(processed_matrix))
```

# Extract transcript files
```{r, eval = FALSE}
transcript_counts <- list.map(
    seq(3),
    f(i) ~ {
        list.map(
            list.files(
                path = file.path(path, "RNASeq", "transcript_counts", paste0("Batch", i)),
                full.names = TRUE,
                pattern = "tsv"
            ),
            f(i) ~ {
                read_tsv(i) %>%
                    select(target_id, est_counts) %>%
                    set_colnames(c("id", str_extract(i, "\\d{5}[AB]05PB1")))
            }
        ) %>% Reduce(function(x, y) full_join(x, y, by = "id"), .)
    }
) %>% Reduce(function(x, y) full_join(x, y, by = "id"), .)
# use_data(transcript_counts, overwrite = TRUE)
```

# Extract kallisto files
```{r, eval = FALSE}
kallisto_files0 <- list.map(
    c("First", "Second", "Third"),
    f(i) ~ {
        list.map(
            list.files(
                path = file.path(path, "kallisto", paste0(i, "Batch_h5files")),
                full.names = TRUE,
                pattern = "h5"
            ), f(i) ~ {
                tmp <- H5Fopen(i)
                # h5closeAll()
                data.frame(tmp$est_counts) %>% set_colnames(c(str_extract(i, "\\d{5}[AB]05PB1")))
            }
        ) %>% Reduce(cbind, .)
    }
) %>% Reduce(cbind, .)
```

# Extract kallisto files
```{r, eval = FALSE}
tr_files <- list.map(
    seq(3),
    f(i) ~ {
        list.files(
            path = file.path(path, "RNASeq", "transcript_counts", paste0("Batch", i)),
            full.names = TRUE,
            pattern = "tsv"
        )
    }
) %>% unlist()
tr_names <- read_tsv(tr_files[1])$target_id %>% str_remove_all("\\.*")
gene_query <- getBM(
    attributes = c("ensembl_transcript_id", "hgnc_symbol"),
    mart = db_annot
)
tximport(., ignoreTxVersion = TRUE, type = "kallisto", txOut = TRUE, tx2gene = tx2gene, ignoreAfterBar = TRUE)
```

# Extract visits and intersect common rows between data
```{r}
# counts_matrix <- transcript_counts
# Split the visits
l_visits <- list.map(
    c("A", "B"),
    f(i) ~ {
        str_detect(colnames(counts_matrix), paste0("^(\\d{5})", i)) %>%
            which() %>%
            c(1, .) %>%
            select(counts_matrix, .)
    }
)

# Formatting the patient names
col_names <- list.map(l_visits, f(.) ~ {
    colnames(.) %>%
        str_remove_all("(A|B)05PB1") %>%
        as.numeric() %>%
        sort()
})

# Select visit A, rename columns, order by patient ID
l_visits <- list.map(
    c("A", "B"),
    f(i) ~ {
        set_colnames(l_visits[[i]], colnames(l_visits[[i]]) %>% str_replace_all(paste0("0?(\\d{4})(A|B)05PB1"), "\\1")) %>%
            select(c("GeneName", as.character(col_names[[i]])))
    }
)
# list.map(c("A", "B"), f(i) ~ l_visits[[i]] %>% select(intersect(colnames(.), rownames(blocks))) %>% ncol() %>% rep(i, .)) %>% unlist() %>% unname() %>% plot_piechart(cex = 30, hsize = 1.5, title = "", label = FALSE, colour = brewer.pal(9, "Paired")[7:8])
counts_matrix <- l_visits[["A"]]
gene_name <- str_remove_all(counts_matrix$GeneName, "\\.*")
# Filter by common rows
common_rows <- Reduce(intersect, list(colnames(counts_matrix), rownames(blocks), names(batches)))
counts_matrix <- select(counts_matrix, -GeneName)

blocks <- data.frame(blocks)[common_rows, ] %>% select(contains(c("Mucocutaneous", "Musculoskeletal", "Urticaria", "Myocarditis", "Anti.IL1", "Anti.IL6", "Corticosteroid", "DMARDs", "Kinase.Inhibitor", "Naive")))

# c(list.map(l_visits, f(i) ~ colnames(i)[-1]), CLINIC = list(rownames(blocks)))
# plot_venn1(temp, color = c("#FED976", "#FD8D3C", "#377EB8"))
# blocks %>% mutate(Biologic = Anti.IL1 + Anti.IL6) %>% select(c(3:5, 7)) %>% plot_bar_mcat(colors = get_colors()[c(seq(3), 5)], title = "") -> p
# select(blocks, seq(2)) %>% plot_bar_mcat(colors = brewer_pal(, "YlGn")(9)[c(4, 8)], title = "") -> p2
# blocks0 %>% select(c(1, 3)) %>% mutate(id = rownames(.)) -> df
# blocks %>% select(c(2:3)) %>% mutate(id = rownames(.)) -> df2
```
# Reshape data for treatments
```{r}
blocks0 <- blocks
if ("Corticosteroid" %in% colnames(blocks)) {
    selected_comb <- c("Naive vs GCs only", "Naive vs aIL1 only", "aIL1 only vs GCs + aIL1", "Naive vs GCs + aIL1", "GCs only vs GCs + aIL1")
    blocks <- blocks %>%
        mutate(`Naive vs GCs only` = (Corticosteroid == 1 & Anti.IL1 != 1 & Anti.IL6 != 1 & DMARDs != 1) | Naive == 1, `Naive vs aIL1 only` = (Corticosteroid != 1 & Anti.IL1 == 1 & Anti.IL6 != 1 & DMARDs != 1) | Naive == 1, `GCs only vs GCs + aIL1` = (Corticosteroid == 1 & Anti.IL1 != 1 & Anti.IL6 != 1 & DMARDs != 1) | (Corticosteroid == 1 & Anti.IL1 == 1 & Anti.IL6 != 1 & DMARDs != 1), `aIL1 only vs GCs + aIL1` = (Corticosteroid != 1 & Anti.IL1 == 1 & Anti.IL6 != 1 & DMARDs != 1) | (Corticosteroid == 1 & Anti.IL1 == 1 & Anti.IL6 != 1 & DMARDs != 1), `Naive vs GCs + aIL1` = (Corticosteroid == 1 & Anti.IL1 == 1 & Anti.IL6 != 1 & DMARDs != 1) | Naive == 1, Anti.IL12 = Anti.IL1, Corticosteroid2 = Corticosteroid) %>%
        select(Naive, Anti.IL1, Corticosteroid, Corticosteroid2, Anti.IL12, selected_comb)
    blocks <- list.map(seq(length(blocks) / 2), f(i) ~ filter(blocks, blocks[, i + length(blocks) / 2] == TRUE) %>% select(i)) %>% setNames(selected_comb)
    blocks[[1]] <- mutate(blocks[[1]], Naive = ifelse(Naive == 1, 0, 1))
    # blocks[[2]] <- mutate(blocks[[2]], Anti.IL1 = ifelse(Anti.IL1 == 1, 0, 1))
} else {
    blocks <- list.map(colnames(blocks), f(i) ~ select(blocks, i))
}
```

# Annotation query
```{r, eval = FALSE}
db_annot <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl", version = 104)
# datasets <- listDatasets(db_annot)
# filter(datasets, str_detect(description, "Human")) %>% pull(version)

gene_query <- getBM(
    attributes = c("ensembl_gene_id", "hgnc_symbol"),
    filters = c("ensembl_gene_id", "transcript_biotype", "chromosome_name"),
    values = list(gene_name, "protein_coding", seq(22)),
    mart = db_annot
)

# gene_query <- read_tsv(file = file.path(path, "martquery.tsv")) %>% set_colnames(c("ensembl_gene_id", "hgnc_symbol"))
# which(is.na(gene_query$`Gene name`)) %>% length()
gene_query[duplicated(gene_query$ensembl_gene_id) %>% which(), ] %>%
    pull(ensembl_gene_id) %>%
    paste0("^", ., "$") %>%
    list.map(., f(i) ~ filter(gene_query, str_detect(ensembl_gene_id, i)))

duplicates <- gene_query[duplicated(gene_query$ensembl_gene_id) %>% which(), 1]
gene_query <- gene_query[!gene_query$ensembl_gene_id %in% duplicates, ] %>% rbind(data.frame(ensembl_gene_id = c("ENSG00000276085"), hgnc_symbol = c("CCL3L1/3")))
use_data(gene_query, overwrite = TRUE)
```

# Join gene counts with the annotation
```{r}
data(gene_query)
counts_matrix <- mutate(counts_matrix, ensembl_gene_id = gene_name) %>%
    left_join(gene_query) %>%
    arrange(hgnc_symbol) %>%
    # select(-ensembl_gene_id) %>%
    data.frame() %>%
    filter(!str_detect(hgnc_symbol, "^$"))
gene_hgnc <- pull(counts_matrix, hgnc_symbol) %>%
    snakecase::to_any_case(case = "none", sep_in = NULL, unique_sep = "_")
counts_matrix <- select(counts_matrix, -c(hgnc_symbol)) %>%
    set_colnames(colnames(counts_matrix)[-c(ncol(counts_matrix))] %>% str_remove_all("^X")) %>%
    set_rownames(gene_hgnc)
# use_data(counts_matrix, overwrite = TRUE)
```

# Extract gene files
```{r, eval = FALSE}
gene_counts <- list.map(
    seq(3),
    f(i) ~ {
        list.map(
            list.files(
                path = file.path(path, "RNASeq", "gene_counts", paste0("Batch", i)),
                full.names = TRUE,
                pattern = "tsv"
            ),
            f(i) ~ {
                read_tsv(i, col_names = FALSE) %>%
                    select(X1, X2) %>%
                    set_colnames(c("id", str_extract(i, "\\d{5}[AB]05PB1")))
            }
        ) %>% Reduce(function(x, y) full_join(x, y, by = "id"), .)
    }
) %>% Reduce(function(x, y) full_join(x, y, by = "id"), .)
# use_data(gene_counts, overwrite = TRUE)
```

# Remove batch effect
```{r, eval = FALSE}
# library(sva)
counts_combat <- as.matrix(counts_matrix) %>%
    set_rownames(gene_name) %>%
    ComBat_seq(batches) %>%
    as.data.frame()
counts_combat_HUGO <- ensembl_to_hugo(ensembl_hgnc, counts_combat)
```

# DGEA
```{r}
# setA <- select(counts_matrix, which(groups == 0) %>% names())
# setB <- select(counts_matrix, which(groups == 1) %>% names())
# comparison <- cbind(setA, setB)
# exprs_threshold <- 1 * NCOL(counts_matrix)
filtered_counts <- counts_matrix[apply(select(counts_matrix, -ensembl_gene_id), 1, sum) > 10, ]
ensembl_ids <- pull(filtered_counts, ensembl_gene_id)
filtered_counts <- select(filtered_counts, -ensembl_gene_id) %>% round()
l_de0 <- list.map(blocks, f(i) ~ {
    DESeqDataSetFromMatrix(
        countData = filtered_counts %>% select(colnames(.) %in% rownames(i) %>% which()),
        colData = data.frame(group = pull(i, 1) %>% factor(), batch = batches[names(batches) %in% rownames(i)] %>% factor()),
        design = ~ batch + group
    ) %>% DESeq()
})
# results() %>%
l_de <- list.map(l_de0, f(i) ~ lfcShrink(i, coef = "group_1_vs_0") %>%
    as.data.frame())
```

# Volcano plot
```{r}
l_de2 <- list.map(l_de, f(i) ~ i %>%
    mutate(ensembl_ids = ensembl_ids, log2fc = log2FoldChange, p = pvalue, log10p = -log(padj, 10), name = rownames(.) %>% str_remove_all("_\\d*") %>% toupper(), Expression = case_when(
        log2FoldChange >= 0.5 & padj <= 0.05 ~ "Up-regulated",
        log2FoldChange <= -0.5 & padj <= 0.05 ~ "Down-regulated",
        TRUE ~ "ns"
    )) %>%
    filter(log2FoldChange >= 0.01 | log2FoldChange <= -0.01)
)

list.map(l_de2, f(i, k, j) ~ i %>%
    volcano_plot(cex = 3, legend = "none") + xlab("") + ylab("")) -> p
for (pl in names(p)) {
    save_tiff(p[[pl]], file.path(path_img, paste0("volcano_", pl)), 4000, 3000)
}
top_genes <- list(
    c("IFI44", "MX1", "USP18", "FOXO3", "MCL1", "PTGS2", "VEGFA"),
    c("FAM13A"),
    c("KBTBD11", "TNF", "LIPA", "RAD52", "HCN3"),
    c("FCGR1A", "HIF1A", "ICAM1", "SOD2", "TNFAIP6", "CD3D", "CD3E", "CD3G", "CD8A", "LCK"),
    c("BCL6", "IL10", "VEGFA")
)
list.map(l_de2, f(i, k, j) ~ i %>%
    volcano_plot(top_genes = filter(i, name %in% top_genes[[k]]), cex = 3, legend = "none") + xlab("") + ylab("")) -> p2
for (pl in names(p2)) {
    save_tiff(p2[[pl]], file.path(path_img, paste0("volcano2_", pl)), 4000, 3000)
}

# list(pluck(., 1), pluck(., 2), NULL, pluck(., 3), pluck(., 4)) %>%
# list(pluck(., 1), NULL, NULL, pluck(., 2)) %>%
plot_grid(
    plotlist = p[c(seq(2), 4, 5, 3)],
    nrow = 2,
    ncol = 3,
    align = "hv"
) %>%
    save_tiff(file.path(path_img, "volcano_treat2"), 8000, 4000)
```

# Descriptive statistics
```{r, eval = FALSE}
get_top(l_de2[[5]], 2, 0.05, n = 200) %>%
    mutate(rank_p = rank(desc(log10p)), rank_fc = rank(desc(abs(log2FoldChange)))) %>%
    mutate(rank = rank(rank_p + rank_fc)) %>%
    arrange(rank)

list.map(l_de, f(i) ~ filter(i, !is.na(padj)) %>%
    filter(log2FoldChange >= 1 | log2FoldChange <= -1) %>%
    filter(padj <= 0.05) %>%
    arrange(padj))

de_gene_names <- list.map(l_de, f(i) ~ mutate(i, log2fc = log2FoldChange) %>%
    filter(!is.na(padj)) %>%
    filter(!str_detect(rownames(.), "unknown")) %>%
    get_top(0.5, 1) %>%
    rownames() %>%
    str_remove_all("_\\d") %>%
    unique() %>%
    toupper())

res <- results(l_de0[[4]])
summary(res)
as.data.frame(res) %>%
    arrange(log2FoldChange) %>%
    head()

without_unknown <- res %>%
    as.data.frame() %>%
    filter(!str_detect(rownames(.), "unknown"))
(de_genes <- res %>% as.data.frame() %>% arrange(padj) %>% filter(padj <= 0.05))

de_gene_names <- de_genes %>%
    rownames() %>%
    str_remove_all("_\\d") %>%
    unique() %>%
    toupper()
de_genes_ranked <- without_unknown$log2FoldChange %>%
    set_names(rownames(without_unknown) %>% str_remove_all("_\\d*") %>% toupper()) %>%
    sort(decreasing = TRUE)
# plotMA(res, alpha = 0.05, colSig = "red", ylim = 3 * c(-1, 1))
res %>%
    as.data.frame() %>%
    arrange(desc(log2FoldChange)) %>%
    head()
i <- 2
plot_count <- function(i, gene = gene, cex = 1.5) {
    name <- str_split(names(blocks)[i], " vs ")[[1]]
    plotCounts(l_de0[[i]], gene = gene, intgroup = "group", returnData = TRUE) %>%
    mutate(group = factor(group, labels = c(name[1], name[2]))) %>%
    ggplot(aes(x = group, y = count, color = group)) +
    geom_point(position = position_jitter(w = 0.1, h = 0), size = cex * 3, alpha = 0.5) +
    geom_text_repel(aes(label = colnames(l_de0[[i]]), x = group, y = count, color = group), size = cex * 3) +
    scale_y_log10() +
    theme_bw() +
    labs(
        y = "Count",
        x = gene
    ) +
    scale_color_manual(values = get_colors()[c(2, 4)], name = gene) +
    theme(
        legend.position = "none",
        axis.text = element_text(colour = "gray50", size = cex * 10),
        axis.title = element_text(colour = "gray50", size = cex * 12, face = "bold.italic"),
        axis.ticks = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.border = element_blank()
    )
}
plot_count(i, "VEGFA")

list.map(top_genes[[i]], f(x) ~ plot_count(i, gene = x) + ylab("")) %>% plot_grid(
    plotlist = .,
    nrow = 3,
    ncol = 3,
    align = "hv"
)
# use_data(l_de_manif, overwrite = TRUE)

outs <- rownames(blocks[[i]]) %>% paste0("^", ., "$", collapse = "|")
name <- str_split(names(blocks)[i], " vs ")[[1]]
attributes(clinic_intersect)$code %>% filter(str_detect(column_code, paste0("^", c("C_2643_3796", "C_2643_3803", "C_2649_3888", "C_2649_3895", "C_2649_3891", "C_2643_3797"), "$", collapse = "|"))) %>% pull(item_name) %>% c("batch") %>% select(clinic_intersect, .) %>% mutate(clinical_center = str_remove_all(clinical_center, "\\d{2} - ") %>% str_remove_all(" \\(.+\\)") %>% str_remove_all(".+ [-â€“] ")) %>% filter(str_detect(immun_aid_identifier, outs)) %>% mutate(Group = ifelse(blocks[[i]] == 0, name[1], name[2])) %>% select(2, 8, 3:4, 7, 6, 1) %>% set_colnames(colnames(.) %>% str_replace_all("_", " ") %>% str_to_sentence()) %>% arrange(Group) %>% kable0()


list.map(list("IL10", "BCL6", "VEGFA"), f(x) ~ plotCounts(l_de0[[i]], gene = x, intgroup = "group", returnData = TRUE) %>% select(-group) %>% set_colnames(x)) %>% Reduce(cbind, .)
```

# Significant genes
```{r}
de_genes <- list.map(l_de2, f(i) ~ get_top0(i, 0.5, 0.05, n = 200) %>%
    rownames() %>%
    str_remove_all("_\\d") %>%
    unique())
cbind(sapply(blocks, function(i) paste0(sum(i), "/", which(i == 0) %>% length())) %>% as.data.frame() %>% set_colnames(c("N. patients")), sapply(de_genes, length) %>% as.data.frame() %>% set_colnames(c("N. significant genes"))) %>% kable0()

# fc_cut <- c(1.5, 1.5)
fc_cut <- c(0.5, 1.5)
de_genes1 <- list.map(l_de2, f(i) ~ {
    list.map(c("Up-regulated", "Down-regulated"), f(j, k) ~ {
        get_top0(i, 1, 0.05) %>%
            filter(Expression == j) %>%
            rownames() %>%
            str_remove_all("_\\d") %>%
            unique()
    })
})

sapply(de_genes, function(x) paste(sapply(x, length), sep = "/"))

cbind(sapply(blocks, function(i) paste0(which(i == 0) %>% length(), "/", sum(i))) %>% as.data.frame() %>% set_colnames(c("N. patients")), sapply(de_genes1, function(x) paste(sapply(x, length), collapse = "/")) %>% as.data.frame() %>% set_colnames(c("N. significant genes (Up-/Down-regulated)"))) %>%
    kable0() %>%
    kable_styling(bootstrap_options = c("striped", "condensed"))
# %>% slice(c(seq(2), 4))

# de_genes_ranked <- list.map(l_de, f(i) ~ {filter(i, !is.na(padj)) %>% filter(log2FoldChange >= 0.5 | log2FoldChange <= -0.5) %>% filter(pvalue <= 0.2) %>% filter(!str_detect(rownames(.), "Unknown")) -> tmp; tmp$log2FoldChange %>%
#     set_names(rownames(tmp) %>% str_remove_all("_\\d*")) %>%
#     sort(decreasing = TRUE)})

de_genes_ranked <- list.map(l_de2, f(i, j) ~ {
        get_top0(i, 0.01, 1, n = 500, rank = TRUE) -> tmp
        tmp$pfc %>%
        set_names(tmp$ensembl_ids) %>%
        sort(decreasing = TRUE)
})
sapply(de_genes_ranked, length)
```

# Enrichment analysis
```{r}
enriched <- list.map(compact(de_genes1), f(i) ~ {
    x <- i[[1]]
    if (length(x) > 0) {
        unlist(x) %>%
            head(250) %>%
            enrichr(dbs)
    }
}) %>% compact()
enriched0 <- list.map(compact(de_genes1), f(i) ~ {
    x <- i[[2]]
    if (length(x) > 0) {
        unlist(x) %>%
            head(250) %>%
            enrichr(dbs)
    }
}) %>% compact()
# enriched <- list.map(de_genes, f(i) ~ enrichr(i, dbs))
list.map(enriched, f(x, y, z) ~ {
    keep(x, function(a) nrow(a) > 0) %>%
        list.map(f(i, j) ~ {
            plot_enrich(i, n = 15, title = str_replace_all(dbs[j], "_", " "))
        }) %>%
        plot_grid(
            plotlist = .,
            nrow = 2,
            ncol = 3,
            align = "hv"
        ) %>%
        save_tiff(file.path(path_img, "shrinked", paste0("enrichr_", z, "_skrinked")), 8000, 4000)
})

# x <- 3
# enriched0[[x]] %>%
#     list.map(f(i, j) ~ {
#         plot_enrich(i, n = 15, title = str_replace_all(dbs[j], "_", " "))
#     }) %>%
#     plot_grid(
#         plotlist = .,
#         nrow = 2,
#         ncol = 3,
#         align = "hv"
#     ) %>%
#     save_tiff(file.path(path_img, "shrinked", paste0("enrichr_", names(enriched)[x], "_down3")), 8000, 4000)

# l_enrich <- list.map(de_genes, f(x) ~ enrichr(x, dbs))
# p <- list.map(l_enrich, f(i, j, k) ~ plot_enrich(i[[5]], n = 15, title = str_replace_all(k, "\\.", "-")))
# plot_grid(
#     plotlist = p[c(1, 4)],
#     nrow = 1,
#     ncol = 2,
#     align = "hv"
# )
list.map(enriched, f(i) ~ list.map(i, f(j) ~ filter(j, Adjusted.P.value <= 0.05) %>%
    dplyr::select(c(1, 4, 9)) %>%
    head(15)) %>% keep(function(x) nrow(x) > 0)) %>% compact() -> tmp0
list.map(enriched0, f(i) ~ list.map(i, f(j) ~ filter(j, Adjusted.P.value <= 0.05) %>%
    dplyr::select(c(1, 4, 9)) %>%
    head(15)) %>% keep(function(x) nrow(x) > 0)) %>% compact() -> tmp1

list.map(dbs, f(j) ~ list.filter(tmp0, f(i) ~ !is.null(i[[j]])) %>% list.names())
# list.map(c(1, 3), f(i) ~ tmp[[2]][[i]] %>% plot_enrich(n = 15, cex = 0.85, title = str_replace_all(dbs[i], "_", " ")))
```

# Heatmap
```{r}
blocks1 <- blocks
# blocks2 <- list.map(seq(length(blocks1)), f(i) ~ select(counts_matrix, rownames(blocks[[i]])) %>% filter(rownames(.) %in% head(de_genes0[[i]], 10)))
# blocks2 <- list.map(seq(length(blocks1)), f(i) ~ select(counts_matrix, rownames(blocks[[i]])) %>% filter(rownames(.) %in% sapply(de_genes[[i]], function(x) head(x, 10)) %>% unlist()))
blocks2 <- list.map(top_genes, f(i, j) ~ list.map(i, f(x) ~ plotCounts(l_de0[[j]], gene = x, intgroup = "group", returnData = TRUE) %>% select(-group) %>% set_colnames(x)) %>% Reduce(cbind, .) %>% t()) 
i <- 1
# names_levels <- c("AOSD", "SOJIA")
names_levels <- c("0", "1")
# names_levels <- c("Naive", "GC-Corticoid", "GC-Corticoid + aIL1", "aIL1 only")
# disease <- clinic_intersect %>% filter(rownames(.) %in% colnames(blocks2[[i]])) %>% pull(disease) %>% factor(labels = names_levels)
disease <- ifelse(blocks1[[i]] == 0, names_levels[1], names_levels[2]) %>% set_colnames("Treatment")
row_annotation <- data.frame(Treatment = disease)
# row_annotation[1, ] <- "GC-Corticoid + aIL1"
# row_annotation[2, ] <- "aIL1 only"
res_scaled <- blocks <- t(blocks2[[i]])
res_scaled_na <- scale0(blocks, "minmax")

disease <- ifelse(blocks1[[2]] == 0, names_levels[1], names_levels[2]) %>% set_colnames("Treatment")
row_annotation <- data.frame(Treatment = disease)
blocks_ <- t(blocks2[[2]]) %>%
    data.frame() %>%
    mutate(Treatment = disease) %>%
    set_rownames(paste0(ifelse(disease == 0, "A", "B"), rownames(.))) %>%
    mutate(variable = rownames(.))

select(blocks_, -FAM13A) %>% ggplot(data = ., aes(x = 1, y = as.character(variable), fill = as.numeric(Treatment))) +
    geom_tile() +
    scale_x_continuous(breaks = 1, labels = "Treatment") +
    scale_y_discrete(position = "right") +
    scale_fill_gradientn(colors = get_colors()[c(5, 1)], na.value = "black") +
    theme(
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(), axis.ticks = element_blank(), legend.position = "none", axis.title = element_blank(), axis.text.x = element_text(face = "bold", angle = -45, hjust = 0.25, vjust = 0.25)
    ) +
    theme(axis.text = element_text(colour = "black", size = 14))
select(blocks_, -Treatment) %>% ggplot(data = ., aes(x = 1, y = as.character(variable), fill = as.numeric(FAM13A))) +
    geom_tile() +
    scale_x_continuous(breaks = 1, labels = rownames(blocks2[[2]])) +
    scale_y_discrete(position = "right") +
    scale_fill_gradientn(colors = colors_ind, na.value = "black") +
    theme(
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(), axis.ticks = element_blank(), legend.position = "none", axis.title = element_blank(), axis.text.x = element_text(angle = -45, hjust = 0.25, vjust = 0.25)
    ) +
    theme(axis.text = element_text(colour = "black", size = 14))
```

# GSEA
```{r}
# ClusterProfiler
organism <- "org.Hs.eg.db"
library(organism, character.only = TRUE)
gse <- list.map(de_genes_ranked, f(i) ~ gseGO(i, ont = "BP", OrgDb = get(organism), keyType = "ENSEMBL"))
# dotplot(gse, showCategory = 10)

list.map(gse, f(i) ~ data.frame(i)) %>%
    keep(function(i) nrow(i) > 0) %>%
    list.map(f(i, j, k) ~ data.frame(i) %>%
        mutate(Adjusted.P.value = p.adjust, Term = str_to_title(Description), Count = setSize, Overlap = core_enrichment) %>%
        plot_enrich(15, gsea = TRUE, title = k)) %>%
    # list(pluck(., 1), pluck(., 2), NULL, pluck(., 3), pluck(., 4)) %>%
    plot_grid(
        plotlist = .,
        nrow = 2,
        ncol = 3,
        align = "hv"
    ) %>% save_tiff(file.path(path_img, "gsea_treat"), 8000, 4000)

enrichplot::pairwise_termsim(gse[[1]]) %>% emapplot(showCategory = 10)
cnetplot(gse[[1]], categorySize="pvalue", foldChange=de_genes_ranked[[1]], showCategory = 3)
gseaplot(gse[[1]], by = "all", title = gse$Description[1], geneSetID = 1)
ridgeplot(gse[[1]]) + labs(x = "enrichment distribution")
```

# Kegg pathways
```{r, eval = FALSE}
c("CD3D", "CD3E", "CD3G", "CD8A", "LCK") %>% bitr(fromType = "ALIAS", toType = c("ENTREZID"), OrgDb = get(organism)) -> tmp
kegg_genes <- list.map(de_genes_ranked, f(i) ~ {
    bitr(names(i), fromType = "ENSEMBL", toType = c("ENTREZID"), OrgDb = get(organism)) %>% left_join(data.frame(ENSEMBL = names(i), log2FoldChange = i)) -> tmp
    tmp$log2FoldChange %>%
        set_names(pull(tmp, ENTREZID)) %>%
        sort(decreasing = TRUE)
})
sapply(kegg_genes, length)

tmp <- list.map(kegg_genes, f(i) ~ gseKEGG(
    geneList = i,
    organism = "hsa",
    keyType = "ncbi-geneid"
))
```
