---
title: "QC_Counts"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{QC_Counts}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
path <- file.path(golem::get_golem_wd(), "inst", "extdata")
gene_counts <- read_tsv(file.path(path, "gene_counts_batch4.tsv")) %>%
  dplyr::rename(ensembl_gene_id = gene_id)
gene_counts_norm <- round(gene_counts[, -1]) %>%
  as.matrix() %>%
  vst() %>%
  as_tibble() %>%
  mutate(ensembl_gene_id = gene_counts$ensembl_gene_id) %>%
  relocate(ensembl_gene_id)

# db_annot <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl", version = 104)
# gene_query <- getBM(
#   attributes = c("ensembl_gene_id", "hgnc_symbol", "chromosome_name", "transcript_biotype", "description", "wikigene_description"),
#   filters = c("ensembl_gene_id", "chromosome_name"),
#   values = list(pull(gene_counts, 1), c(seq(22), "MT", "Y", "X")),
#   mart = db_annot
# )
# write_tsv(gene_query, file = file.path(path, "gene_query.tsv"))
gene_query <- read_tsv(file.path(path, "gene_query.tsv"))

metadata <- inner_join(gene_counts, gene_query, by = "ensembl_gene_id") %>% mutate(chromosome_name = gsub("^(\\d{1})$", "0\\1", chromosome_name))

x <- "chromosome_name"; filter(metadata, !str_detect(!!sym(x), "CHR_")) %>% pull(!!sym(x)) %>% unique()

df0 <- pivot_longer(metadata, cols = ends_with("05PB1"), names_to = "Sample", values_to = "Count")
p <- ggplot(df0, aes(x = as.factor(chromosome_name), y = Count, color = Sample, group = Sample)) +
  geom_point(position = position_jitter(width = 0.2, height = 0)) +
  # geom_boxplot(width = 0.5, alpha = 0.6, outlier.shape = NA) +
  labs(x = "Chromosome", y = "# Reads") +
  theme_minimal() +
  theme(
    axis.title = element_text(size = 15, face = "bold.italic", color = "gray50"),
    panel.grid.major.x = element_blank(),
    axis.text = element_text(size = 13, color = "gray50"),
    legend.position = "none"
  )
p

ggplot(df0, aes(x = hgnc_symbol, y = Count, color = Sample)) +
  geom_line(size = 1) +
  labs(x = "Counts", y = "Density") +
  theme_minimal() +
  theme(
    axis.title = element_text(size = 15, face = "bold.italic", color = "gray50"),
    panel.grid.major.x = element_blank(),
    axis.text = element_text(size = 13, color = "gray50"),
    legend.position = "none"
  ) +
  scale_color_manual(values = get_colors0()(3)) +
  scale_x_continuous(trans = log10_trans(), breaks = df0$Count) +
  scale_y_continuous(label = label_number(suffix = "%"))
###


plot_dispersion <- function(x) {
  df <- data.frame(Sample = rownames(x), x)
  colnames(df)[-1] <- colnames(x)

  medians <- apply(df[, -1], 2, function(x) median(x[x >= 1], na.rm = TRUE))
  iqrs <- apply(df[, -1], 2, function(x) IQR(x[x >= 1], na.rm = TRUE))

  df0 <- pivot_longer(df, cols = -Sample, names_to = "gene", values_to = "value") %>%
    # filter(value >= 1) %>%
    mutate(value = round(value))
  colors <- get_colors0()(ncol(df) - 1)

  ggplot(df0, aes(x = gene, y = value, color = gene)) +
    geom_pointrange(
      aes(
        y = medians[gene],
        ymin = medians[gene] - 0.5 * iqrs[gene],
        ymax = medians[gene] + 0.5 * iqrs[gene],
        color = gene
      ),
      size = .5,
      shape = 3,
      alpha = .5
    ) +
  # geom_text(aes(label = gene)) +
    geom_point(size = 1, alpha = 0.5) +
    labs(x = "Samples", y = "# Reads", color = "Samples") +
    theme_minimal() +
    theme(
      axis.title = element_text(size = 22, face = "bold.italic", color = "gray50"),
      panel.grid.major.x = element_blank(),
      # axis.text.x = element_blank(),
      axis.text.y = element_text(size = 13),
      axis.text.x = element_text(angle = 90, color = colors, size = 10),
      legend.position = "none"
    ) +
    scale_y_continuous(
      # trans = log10_trans(),
      # breaks = c(1, paste0("1e", seq(7))) %>% as.numeric(),
      labels = label_number_auto()
    ) +
    scale_color_manual(values = colors)
    # guides(x = guide_axis(n.dodge = 3))
}

counts0 <- gene_counts[, -1] %>% set_colnames(pull(gene_counts, 1))

counts0 <- t(gene_counts0[, -1]) %>% set_colnames(pull(gene_counts0, 1) %>% str_remove_all("ENSG"))

path_tsv <- file.path(golem::get_golem_wd(), "inst", "results", "tsv")
png(
  file.path(path_tsv, paste0("dispersion", ".temp.", "png")),
  units = "px",
  width = 1200*4,
  height = 575*4,
  res = 300
)
gene_counts_norm[, -1] %>% set_colnames(colnames(.) %>% str_remove_all("05PB1")) %>% plot_dispersion()
dev.off()

apply(gene_counts[, -1] > 0, 2, sum) %>% `/`(nrow(gene_counts)) %>% plot_violin0(title = "") + scale_y_continuous(labels = label_percent())
```
